conflictlog/pool.go

```diff
@@ -2,82 +2,248 @@ package conflictlog
+
+	conn = clist.pop()
+	return
+}
+
+// UpdateGCInterval updates the new GC frequency
+func (pool *connPool) UpdateGCInterval(d time.Duration) {
+	// Negative duration will cause the timer to panic.
+	if d <= 0 {
+		return
 	}
-	defer pool.poolLock.Unlock()
 
-	return wlist
+	pool.mu.Lock()
+	pool.gcTicker = time.NewTicker(d)
+	pool.mu.Unlock()
 }
 
-func (pool *writerPool) get(bucketName string) (w Writer, err error) {
-	wlist := pool.getList(bucketName)
+func (pool *connPool) getGCTicker() (t *time.Ticker) {
+	pool.mu.Lock()
+	t = pool.gcTicker
+	pool.mu.Unlock()
+
+	return
+}
 
-	w = wlist.pop()
-	if w != nil {
+// UpdateReapInterval updates the reap interval for unused connections
+func (pool *connPool) UpdateReapInterval(d time.Duration) {
+	// Negative duration does not make any sense
+	if d <= 0 {
 		return
 	}
 
-	w, err = pool.createWriterFn(bucketName)
-	if err != nil {
+	pool.mu.Lock()
+	pool.reapInterval = d
+	pool.mu.Unlock()
+}
+
+// Get returns an object from the pool. If there is none then it creates
+// one by calling newConnFn() and returns it.
+func (pool *connPool) Get(bucketName string) (conn io.Closer, err error) {
+	conn = pool.get(bucketName)
+	if conn != nil {
 		return
 	}
 
+	conn, err = pool.newConnFn(bucketName)
 	return
 }
 
-func (pool *writerPool) release(w Writer) {
-	wlist := pool.getList(w.Bucket())
-	wlist.push(w)
+func (pool *connPool) Put(bucketName string, conn io.Closer) {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	l := pool.getOrCreateListNoLock(bucketName)
+	l.push(conn)
+}
+
+func (pool *connPool) Close() {
+
+}
+
+// reapConnList collects all the connection lists which have not been used for the reapInterval time
+// This does not close the connection itself
+func (pool *connPool) reapConnList() []*connList {
+	connListList := []*connList{}
+	reapedBuckets := []string{}
+
+	now := time.Now()
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	// Collect all expired buckets and its lists
+	for bucketName, connList := range pool.buckets {
+		elapsed := now.Sub(connList.lastUsed)
+		if elapsed >= pool.reapInterval {
+			reapedBuckets = append(reapedBuckets, bucketName)
+			connListList = append(connListList, connList)
+		}
+	}
+
+	// Delete expired buckets
+	// It is possible that a connection for a deleted bucket will be requested
+	// after this. This is fine since it will treated like a new bucket being requested
+	// for the first time
+	for _, bucketName := range reapedBuckets {
+		pool.logger.Debugf("reaping connections for bucket=%s", bucketName)
+		delete(pool.buckets, bucketName)
+	}
+
+	return connListList
+}
+
+// gcOnce runs one single iteration of reaping the connections
+func (pool *connPool) gcOnce() {
+	connListList := pool.reapConnList()
+
+	// Note: the closing of the connections happen outside the pool lock.
+	// From this point, a parallel request to create a connection is safe
+	// and it will land in a new connList.
+	for _, connList := range connListList {
+		connList.closeAll()
```

Comment:
Assuming closeAll() will actually do connection close on a per list connList basis, is it worth doing this in parallel + with a timeout that bails?